File: backend/app/routers/exams.py
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_current_user, get_db
from app.models import Exam, ExamSession, Provider, User
from app.schemas.exam import ExamDetailResponse, ExamResponse
from app.services.progress_service import get_progress_summary

router = APIRouter(prefix="/exams", tags=["exams"])


@router.get("", response_model=List[ExamResponse])
async def list_exams(provider_id: Optional[int] = None, db: AsyncSession = Depends(get_db)):
    query = select(Exam, Provider.name).join(Provider, Exam.provider_id == Provider.id)
    if provider_id:
        query = query.where(Exam.provider_id == provider_id)
    query = query.order_by(Exam.code)

    result = await db.execute(query)
    return [
        ExamResponse(
            id=exam.id,
            provider_id=exam.provider_id,
            code=exam.code,
            name=exam.name,
            description=exam.description,
            total_questions=exam.total_questions,
            is_active=exam.is_active,
            provider_name=provider_name,
            num_questions=exam.num_questions,
            pass_percentage=exam.pass_percentage,
            time_limit_minutes=exam.time_limit_minutes,
        )
        for exam, provider_name in result.all()
    ]


@router.get("/{exam_id}", response_model=ExamDetailResponse)
async def get_exam(
    exam_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Exam, Provider.name).join(Provider, Exam.provider_id == Provider.id).where(Exam.id == exam_id)
    )
    row = result.one_or_none()
    if not row:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Exam not found")

    exam, provider_name = row
    summary = await get_progress_summary(db, exam_id, user.id)

    # Check for active exam session
    active_result = await db.execute(
        select(ExamSession.id).where(
            ExamSession.user_id == user.id,
            ExamSession.exam_id == exam_id,
            ExamSession.status == "in_progress",
        )
    )
    active_row = active_result.scalar_one_or_none()

    return ExamDetailResponse(
        id=exam.id,
        provider_id=exam.provider_id,
        code=exam.code,
        name=exam.name,
        description=exam.description,
        total_questions=exam.total_questions,
        is_active=exam.is_active,
        provider_name=provider_name,
        num_questions=exam.num_questions,
        pass_percentage=exam.pass_percentage,
        time_limit_minutes=exam.time_limit_minutes,
        progress_summary=summary,
        active_session_id=active_row,
    )

File: workers-api/src/routes/exams.ts
import { Hono } from "hono";
import { eq, and } from "drizzle-orm";
import type { AppEnv } from "../types/env";
import { exams, providers, examSessions } from "../db/schema";
import { getProgressSummary } from "../services/progress-service";
import { AppError } from "../lib/errors";

const examRoutes = new Hono<AppEnv>();

// GET /exams — list (optionally by provider_id)
examRoutes.get("/", async (c) => {
  const db = c.get("db");
  const providerId = c.req.query("provider_id");

  let query = db
    .select({
      id: exams.id,
      providerId: exams.providerId,
      code: exams.code,
      name: exams.name,
      description: exams.description,
      totalQuestions: exams.totalQuestions,
      isActive: exams.isActive,
      numQuestions: exams.numQuestions,
      passPercentage: exams.passPercentage,
      timeLimitMinutes: exams.timeLimitMinutes,
      providerName: providers.name,
    })
    .from(exams)
    .innerJoin(providers, eq(exams.providerId, providers.id))
    .orderBy(exams.code);

  const rows = providerId
    ? await query.where(eq(exams.providerId, Number(providerId)))
    : await query;

  return c.json(
    rows.map((r) => ({
      id: r.id,
      provider_id: r.providerId,
      code: r.code,
      name: r.name,
      description: r.description,
      total_questions: r.totalQuestions,
      is_active: r.isActive,
      num_questions: r.numQuestions,
      pass_percentage: r.passPercentage,
      time_limit_minutes: r.timeLimitMinutes,
      provider_name: r.providerName,
    }))
  );
});

// GET /exams/:id — detail with progress summary + active session
examRoutes.get("/:id", async (c) => {
  const db = c.get("db");
  const user = c.get("user");
  const examId = Number(c.req.param("id"));

  const rows = await db
    .select({
      id: exams.id,
      providerId: exams.providerId,
      code: exams.code,
      name: exams.name,
      description: exams.description,
      totalQuestions: exams.totalQuestions,
      isActive: exams.isActive,
      numQuestions: exams.numQuestions,
      passPercentage: exams.passPercentage,
      timeLimitMinutes: exams.timeLimitMinutes,
      providerName: providers.name,
    })
    .from(exams)
    .innerJoin(providers, eq(exams.providerId, providers.id))
    .where(eq(exams.id, examId));

  if (rows.length === 0) throw new AppError(404, "Exam not found");
  const exam = rows[0];

  const progressSummary = await getProgressSummary(db, examId, user.id);

  // Find active session
  const activeSession = await db.query.examSessions.findFirst({
    where: and(
      eq(examSessions.userId, user.id),
      eq(examSessions.examId, examId),
      eq(examSessions.status, "in_progress")
    ),
  });

  return c.json({
    id: exam.id,
    provider_id: exam.providerId,
    code: exam.code,
    name: exam.name,
    description: exam.description,
    total_questions: exam.totalQuestions,
    is_active: exam.isActive,
    num_questions: exam.numQuestions,
    pass_percentage: exam.passPercentage,
    time_limit_minutes: exam.timeLimitMinutes,
    provider_name: exam.providerName,
    progress_summary: progressSummary,
    active_session_id: activeSession?.id ?? null,
  });
});

export default examRoutes;
